CRM/Core/BAO/MessageTemplate.php

```diff
@@ -123,7 +123,8 @@ public static function add(&$params) {
       case 'idname':
         $map = self::getWorkflowNameIdMap();
         if ($map[$params['workflow_name']] != $params['workflow_id']) {
-          throw new CRM_Core_Exception("The workflow_id and workflow_name are mismatched. Note: You only need to submit one or the other.");
+          unset($params['workflow_id']);
+          Civi::log()->info('The workflow_id and workflow_name are mismatched. Note: You only need to submit one or the other.');
```

Comment:
I was a bit befuddled at how one would get into this code-path -- in the use-case you describe,  there is no `workflow_id`. The `switch()` statement is supposed to do direct that use-case down the other branch (`case 'name':`)

The problem is `null`/`"null"` handling. You submit an API call with `['workflow_id'=>null]`, and it gets [internally rewritten](https://github.com/civicrm/civicrm-core/blob/5.42/Civi/Api4/Utils/FormattingUtil.php#L57) to `['workflow_id'=>'null']` 🤦. Then the guard for `empty($params['workflow_id'])` determines "Not Empty!" for both id+name. So it tries to match them.

Note that converting this to `log()->info()` isn't sufficient on two grounds: (1) it creates noise in the otherwise reasonable use-case that you described and (2) the older name/id mappings will be more breaky.

There is the question of whether `workflow_id` is still expected to work (ie for older workflows/callers), ie:

* If "No", then rip it out.
* If "Yes", then fix it.

Reading between the lines on this patch,it seems to have taken the conservative approach - ie it doesn't try to rip-out that mechanism. (And that seems better for avoiding a game of whackamole.) So I've opened an alternative #21694.

Reply:
@totten yeah I guess what we tend to do is each time we revisit something we remove it a bit more - the risk is that sometimes it turns out not to be clear that code / fields are on the way out & they get written back in